hazelcast-client/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheInvalidationTest.java

```diff
@@ -0,0 +1,336 @@
+        NearCacheManager nearCacheManager = client.client.getNearCacheManager();
+        CachingProvider provider = HazelcastClientCachingProvider.createCachingProvider(client);
+        HazelcastClientCacheManager cacheManager = (HazelcastClientCacheManager) provider.getCacheManager();
+        HazelcastServerCacheManager memberCacheManager = (HazelcastServerCacheManager) HazelcastServerCachingProvider
+                .createCachingProvider(member).getCacheManager();
+
+        ICache<Object, String> cache = cacheManager.createCache(DEFAULT_CACHE_NAME, createCacheConfig(inMemoryFormat));
+        ICache<Object, String> memberCache = member.getCacheManager().getCache(
+                CacheUtil.getPrefixedCacheName(DEFAULT_CACHE_NAME, null, null));
+
+        NearCache<Data, String> nearCache = nearCacheManager.getNearCache(
+                cacheManager.getCacheNameWithPrefix(DEFAULT_CACHE_NAME));
+
+        testContext = new NearCacheTestContext(client, member, cacheManager, memberCacheManager, nearCacheManager, cache,
+                memberCache, nearCache);
+
+        // add 1 K-V pair in the cache
+        testContext.cache.put("a", "b");
+        testContext.cache.get("a");
+
+    }
+
+    @After
+    public void tearDown() {
+        hazelcastFactory.shutdownAll();
+    }
+
+    @Test
+    public void when_cacheDestroyed_invalidationEventIsReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        destroy();
+        assertInvalidationEventReceivedExactlyOnce(counter);
+    }
+
+    @Test
+    public void when_cacheCleared_invalidationEventIsReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        clear();
+        assertInvalidationEventReceivedExactlyOnce(counter);
+    }
+
+    @Test
+    public void when_cacheClosed_invalidationEventIsNotReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        close();
+        assertInvalidationEventNeverReceived(counter);
+    }
+
+    // When CacheManager.destroyCache is invoked from client-side CacheManager, an invalidation event is received
+    // When invoked from a member-side CacheManager, invalidation event is not received.
+    @Test
+    public void when_cacheManagerDestroyCacheInvoked_invalidationEventMayBeReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        destroyCacheFromCacheManager();
+        if (invokeCacheOperationsFromMember) {
+            // when from a member-side cache manager, invalidation event cannot be received, as the invalidation listener
+            // registration is removed *before* the invocation to destroy the cache is sent to the member
+            assertInvalidationEventNeverReceived(counter);
```

Comment:
Isn't this a problem for working of client near cache? Client near cache will never be invalidated in this case and it will have stale values.